### PR TITLE
Immutable tag api

### PR DESCRIPTION
**Pull request by stonezdj**
Add immutable tags API
Signed-off-by: stonezdj <stonezdj@gmail.com>

### CODE REVIEW

File: src/core/api/immutabletagrule.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+func (itr *ImmutableTagRuleAPI) requireAccess(action rbac.Action) bool {
+	return itr.RequireProjectAccess(itr.projectID, action, rbac.ResourceImmutableTag)
+}
+
+// List list all immutable tag rules of current project
+func (itr *ImmutableTagRuleAPI) List() {
+	rules, err := itr.manager.QueryImmutableRuleByProjectID(itr.projectID)
+	if err != nil {
+		itr.SendInternalServerError(err)
```
Comment (wy65701436) [inline src/core/api/immutabletagrule.go]: return missed?

File: src/core/api/immutabletagrule.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+	if !isValidSelectorJSON(ir.RepoFilter) || !isValidSelectorJSON(ir.TagFilter) {
+		itr.SendBadRequestError(fmt.Errorf("the filter should be a valid json"))
+		return
+	}
+
+	ir.ProjectID = itr.projectID
+	id, err := itr.manager.CreateImmutableRule(ir)
+	if err != nil {
+		itr.SendInternalServerError(err)
```
Comment (wy65701436) [inline src/core/api/immutabletagrule.go]: ditto

File: src/core/api/immutabletagrule.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	}
+	itr.Redirect(http.StatusCreated, strconv.FormatInt(id, 10))
+
+}
+
+// Delete delete immutable tag rule
+func (itr *ImmutableTagRuleAPI) Delete() {
+	if itr.ID <= 0 {
+		itr.SendBadRequestError(fmt.Errorf("invalid immutable rule id %d", itr.ID))
```
Comment (wy65701436) [inline src/core/api/immutabletagrule.go]: ditto

File: src/core/api/immutabletagrule.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	ir := &models.ImmutableRule{}
+	if err := itr.DecodeJSONReq(ir); err != nil {
+		itr.SendInternalServerError(err)
+		return
+	}
+	ir.ID = itr.ID
+	ir.ProjectID = itr.projectID
+
+	if itr.ID <= 0 {
+		itr.SendBadRequestError(fmt.Errorf("invalid immutable rule id %d", itr.ID))
```
Comment (wy65701436) [inline src/core/api/immutabletagrule.go]: ditto

File: src/core/api/immutabletagrule.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	itr.projectID = pid
+
+	ruleID, err := itr.GetInt64FromPath(":id")
+	if err == nil || ruleID > 0 {
+		itr.ID = ruleID
+	}
+
+	itr.manager = immutabletag.NewDefaultRuleManager()
+
+	if itr.SecurityCtx.IsSysAdmin() {
```
Comment (ywk253100) [inline src/core/api/immutabletagrule.go]: The system admin is already covered in the `requireAccess `

File: src/core/api/immutabletagrule.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+		return
+	}
+
+	if strings.EqualFold(itr.Ctx.Request.Method, "get") {
```
Comment (ywk253100) [inline src/core/api/immutabletagrule.go]: The flow will continue even though the checking fails, you should check the permission in the corresponding methods

File: src/core/api/immutabletagrule.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+		return
+	}
+	itr.WriteJSONData(rules)
+}
+
+// Post create immutable tag rule
+func (itr *ImmutableTagRuleAPI) Post() {
+	ir := &models.ImmutableRule{}
+	if err := itr.DecodeJSONReq(ir); err != nil {
+		itr.SendInternalServerError(err)
```
Comment (ywk253100) [inline src/core/api/immutabletagrule.go]: Should be 400 error

File: src/core/api/immutabletagrule.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+}
+
+// Delete delete immutable tag rule
+func (itr *ImmutableTagRuleAPI) Delete() {
+	if itr.ID <= 0 {
+		itr.SendBadRequestError(fmt.Errorf("invalid immutable rule id %d", itr.ID))
+		return
+	}
+	itr.manager.DeleteImmutableRule(itr.ID)
```
Comment (ywk253100) [inline src/core/api/immutabletagrule.go]: Check the error

Comment (stonezdj) [reply src/core/api/immutabletagrule.go]: Update

File: src/core/api/immutabletagrule.go (hunk 7 above)
Comment (stonezdj) [reply src/core/api/immutabletagrule.go]: Update

File: src/core/api/immutabletagrule.go (hunk 5 above)
Comment (stonezdj) [reply src/core/api/immutabletagrule.go]: not covered

**Pull request merged**